src/functional/fencil_processors/runners/gtfn_cpu.py

```diff
@@ -13,36 +13,56 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 
-from typing import Any
+from dataclasses import dataclass, field
+from typing import Any, Optional
 
-import numpy
+import numpy as np
 
-from functional.fencil_processors.builders import cpp as cpp_callable
-from functional.fencil_processors.codegens.gtfn import gtfn_module as gtfn_codegen
-from functional.fencil_processors.processor_interface import fencil_executor
-from functional.iterator import ir
+from functional.fencil_processors import processor_interface as fpi  # fencil processor interface
+from functional.fencil_processors.builders import cache
+from functional.fencil_processors.builders.cpp import bindings, build
+from functional.fencil_processors.codegens.gtfn import gtfn_module
+from functional.fencil_processors.source_modules import cpp_gen
+from functional.iterator import ir as itir
 
 
+# TODO(ricoh): Add support for the whole range of arguments that can be passed to a fencil.
 def convert_arg(arg: Any) -> Any:
-    view = numpy.asarray(arg)
+    view = np.asarray(arg)
     if view.ndim > 0:
-        return memoryview(view)
+        return view
     else:
         return arg
 
 
-# TODO(ricoh): change style to declarative pipeline
-@fencil_executor
-def run_gtfn(itir: ir.FencilDefinition, *args, **kwargs):
-    """
-    Execute the iterator IR fencil with the provided arguments.
+@dataclass(frozen=True)
+class GTFNExecutor(fpi.FencilExecutor):
+    language_settings: cpp_gen.CppLanguage = field(default=cpp_gen.CPP_DEFAULT)
+    name: Optional[str] = None
 
-    The fencil is compiled to machine code with C++ as an intermediate step,
-    so the first execution is expected to have a significant overhead, while subsequent
-    calls are very fast. Only scalar and buffer arguments are supported currently.
+    def __call__(self, fencil: itir.FencilDefinition, *args: Any, **kwargs: Any) -> None:
+        """
+        Execute the iterator IR fencil with the provided arguments.
 
-    See ``FencilExecutorFunction`` for details.
-    """
-    source_module = gtfn_codegen.create_source_module(itir, *args, **kwargs)
-    wrapper = cpp_callable.create_callable(source_module)
-    wrapper(*[convert_arg(arg) for arg in args])
+        The fencil is compiled to machine code with C++ as an intermediate step,
+        so the first execution is expected to have a significant overhead, while subsequent
+        calls are very fast. Only scalar and buffer arguments are supported currently.
+
+        See ``FencilExecutorFunction`` for details.
+        """
+        return build.CMakeProject(
```

Comment:
I would like to avoid that. It is an invitation to put custom logic between the steps (rather than improve the steps). If this is not readable enough, we indeed need a pipeline runner of some kind. Something along the lines of

```python
return <pipeline_runner_name>(
    source_module_gen=gtfn_module.GTFNSourceModuleGenerator(self.language_settings),
    binding_module_gen=bindings.create_bindings,
    buildable_project_gen=build.CMakeProject,
    cache_strategy=cache.Strategy.SESSION,
)(fencil).get_implementation()
```

Reply:
I assume you meant something like

```python
# Never ever put any logic in between these lines!
# Instead, if something is missing it has to go into one of the functions we are calling here.
source_module = gtfn_module.GTFNSourceModuleGenerator(self.language_settings)(fencil, *args, **kwargs)
bindings_module = bindings.create_bindings(source_module)
cmake_project =  build.CMakeProject(
    source_module=source_module,
    bindings_module=bindings_module,
    cache_strategy=cache.Strategy.SESSION,
)
cpp_python_extension_function = cmake_project.get_implementation()
cpp_python_extension_function(*[convert_arg(arg) for arg in args])
```

Reply:
Yeah, let's leave it like this with a TODO